lua/orgmode/org/hyperlinks.lua

```diff
@@ -77,98 +43,127 @@ function Hyperlinks.find_by_filepath(ctx)
+  else
     return {}
   end
-  local headlines = Files.get_current_file():find_headlines_by_title(ctx.base, false)
-  if ctx.skip_add_prefix then
-    return headlines
-  end
-  return vim.tbl_map(function(headline)
-    return headline.title
-  end, headlines)
 end
 
-function Hyperlinks.find_matching_links(ctx)
-  ctx = ctx or {}
-  ctx.base = ctx.base and vim.trim(ctx.base) or nil
-
-  update_hyperlink_ctx(ctx)
-
-  if ctx.base:find('^file:') and not ctx.skip_add_prefix then
-    return Hyperlinks.find_by_filepath(ctx)
+---@param url Url
+---@return fun(headlines: Section[]): string[]
+function Hyperlinks.as_dedicated_targets(url)
+  return function(headlines)
+    local targets = {}
+    local term = as_dedicated_anchor_pattern(url:get_dedicated_anchor())
+    for _, headline in ipairs(headlines) do
+      for m in headline.title:lower():gmatch(term) do
+        table.insert(targets, m)
+      end
+      for _, content in ipairs(headline.content) do
+        for m in content:lower():gmatch(term) do
+          table.insert(targets, m)
+        end
+      end
+    end
+    return targets
   end
+end
 
-  local prefix = ctx.base:sub(1, 1)
-  if prefix == '#' then
-    return Hyperlinks.find_by_custom_id_property(ctx)
-  end
-  if prefix == '*' then
-    return Hyperlinks.find_by_title_pointer(ctx)
+---@param url Url
+---@return fun(headlines: Section[]): table<string>
+function Hyperlinks.as_dedicated_anchors_or_internal_titles(url)
+  return function(headlines)
+    local dedicated_anchors = Hyperlinks.as_dedicated_targets(url)(headlines)
+    local fuzzy_titles = Hyperlinks.as_headline_anchors(headlines, true)
+    return utils.concat(dedicated_anchors, fuzzy_titles, true)
   end
-
-  local results = Hyperlinks.find_by_dedicated_target(ctx)
-  local all = utils.concat(results, Hyperlinks.find_by_title(ctx))
-  return all
 end
 
-function Hyperlinks.get_file_real_path(url_path)
-  local path = url_path
-  path = path:gsub('^file:', '')
-  if path:match('^~/') then
-    path = path:gsub('^~', os.getenv('HOME'))
-  end
-  if path:match('^/') then
-    return path
-  end
-  path = path:gsub('^./', '')
-  return vim.fn.fnamemodify(utils.current_file_path(), ':p:h') .. '/' .. path
+---@param url Url
+---@return Section[], fun(headline: Section[]): string[]
+function Hyperlinks.find_matching_links(url)
+  local result = {}
+  local mapper = function(item)
+    return item
+  end
+  if not url then
+    return result, mapper
+  elseif url:is_file_plain() then
+    result = Hyperlinks.find_by_filepath(url)
+  elseif url:is_custom_id() then
+    result = Hyperlinks.find_by_custom_id_property(url)
+    mapper = Hyperlinks.as_custom_id_anchors
+  elseif url:is_headline() then
+    result = Hyperlinks.find_by_title(url)
+    mapper = Hyperlinks.as_headline_anchors
+  elseif url:is_dedicated_anchor_or_internal_title() then
+    result = utils.concat(
+      Hyperlinks.find_by_dedicated_target(url),
+      -- TODO replace with real fuzzy title search
```

Comment:
I'm not sure, if this would be a use case for others, but as a last resort it could be convenient to refer between headlines within a document. Currently it has to match with the beginning of a headline.